### ISSUE TITLE
Support scenario-global batches with task-, dataset-, and GPU-specific microbatches

### ISSUE THREAD

**Issue by jpata** [jpata/particleflow#498]
## Motivation

Training scenarios currently derive one `gpu_batch_multiplier` from a scenario-level global batch, the number of GPUs, and a single physical-dataset batch size. The multiplier directly changes the DataLoader batch and is not gradient accumulation.

This has two limitations:

- Elementwise and set-output tasks have different memory footprints on the same GPU.
- Datasets such as ttbar, WW, and qq can have substantially different event sizes, especially for hit-based training.

Using one per-GPU batch for every task and dataset either wastes accelerator memory or causes OOM failures. We should preserve a scientifically controlled global batch while allowing explicitly benchmarked microbatch capacities.

## Proposed terminology and invariant

Use three explicit concepts:

- `global_batch_size`: number of events contributing to one optimizer update across all ranks; owned by the comparison scenario.
- `per_gpu_batch_size`: simultaneous events processed by one rank; selected from a task/dataset/accelerator capacity table.
- `gradient_accumulation_steps`: derived by the resolver.

The resolver must enforce:

```text
global_batch_size = num_gpus * per_gpu_batch_size * gradient_accumulation_steps
```

Non-integral configurations should fail before submission.

### Scientific scenario

The scenario owns the global batch, comparison variants, and deterministic dataset schedule:

```yaml
variants:
  elementwise:
    model_name: pyg-cld-hits-v1
    batch_profile: cld_hits_attention_elementwise
  set:
    model_name: pyg-cld-hits-set-v1
    batch_profile: cld_hits_attention_set

training:
  global_batch_size: 512
  dataset_schedule:
    mode: weighted_round_robin
    weights:
      ttbar_hits: 1
      ww_fullhad_hits: 1
      qq_hits: 1
```

### Batch-capacity catalog

Keep empirically measured capacities separate from both the scientific scenario and cluster profile. Key entries by workload, GPU memory class, and logical dataset group:

```yaml
batch_profiles:
  cld_hits_attention_set:
    h100_80gb:
      ttbar_hits: {train: 8, eval: 8}
      ww_fullhad_hits: {train: 6, eval: 6}
      qq_hits: {train: 12, eval: 12}
    a100_80gb:
      ttbar_hits: {train: 4, eval: 4}
      ww_fullhad_hits: {train: 3, eval: 3}
      qq_hits: {train: 6, eval: 6}
```

The values above are illustrative and must be replaced by benchmarked capacities. Memory variants such as `a100_40gb` and `a100_80gb` should have distinct keys.

### Platform profile

The platform profile identifies the hardware and allocation, without task-specific batch settings:

```yaml
name: flatiron_h100
accelerator: h100_80gb
gpus: 8
```

## Dataset and scheduling changes

The current CLD hit configuration concatenates ttbar, WW, and qq into one `physical` loader. Split these into stable logical loader groups so each group can use a different microbatch:

```yaml
train_datasets:
  ttbar_hits:
    samples: [cld_edm_ttbar_hits]
  ww_fullhad_hits:
    samples: [cld_edm_ww_fullhad_hits]
  qq_hits:
    samples: [cld_edm_qq_hits]
```

For the initial implementation, each optimizer update should select one dataset group using a deterministic weighted scheduler, then consume enough microbatches from that loader to reach the scenario global batch. This keeps accumulation exact and gives both comparison variants the same dataset sequence and number of events per optimizer update.

## Training-loop semantics

One reported training step must mean one optimizer update:

1. Select the dataset group.
2. Zero gradients.
3. Process the derived number of microbatches.
4. Scale each mean loss by `1 / gradient_accumulation_steps`.
5. Use DDP `no_sync()` for every microbatch except the final one.
6. Step the optimizer, GradScaler, and LR scheduler once.

Consequences:

- `num_steps`, `val_freq`, and `checkpoint_freq` continue to count optimizer updates.
- Checkpoints are written only at accumulation boundaries.
- Training metrics are aggregated across every microbatch in the update.
- Resume state must include the deterministic dataset scheduler state.

Validation and testing do not need to satisfy a global-batch invariant. They should use the explicit `eval` microbatch, defaulting to the training microbatch if omitted.

## Validation and observability

The scenario resolver should print and save a table containing, for every variant and dataset group:

- Accelerator key and GPU count.
- Per-GPU training/evaluation batch.
- Gradient accumulation steps.
- Effective global batch.

Write the resolved table to `scenario-manifest.json` and TensorBoard. Reject missing capacity entries, unknown dataset keys, non-positive values, non-integral accumulation, and different effective global batches between comparison variants.

## Implementation tasks

- [ ] Add `accelerator` identifiers to hardware profiles.
- [ ] Add a validated batch-capacity catalog and variant `batch_profile` references.
- [ ] Split combined physical datasets into logical per-dataset loader groups.
- [ ] Resolve per-dataset microbatch and accumulation settings before launch.
- [ ] Store the fully resolved batch plan in `MLPFConfig` and the scenario manifest.
- [ ] Refactor the training loop into microbatch backward passes plus one optimizer step.
- [ ] Use DDP `no_sync()` during accumulation.
- [ ] Add deterministic optimizer-step-level dataset scheduling and checkpoint state.
- [ ] Aggregate loss, timing, memory, and event-count metrics over microbatches.
- [ ] Support separate evaluation microbatches.
- [ ] Keep `gpu_batch_multiplier` temporarily for legacy non-scenario commands, but stop emitting it from scenario launches.
- [ ] Benchmark elementwise/set workloads for each dataset on H100 80 GB, A100 variants, B200, and supported local GPUs.

## Acceptance criteria

- Elementwise and set variants use the same configured global batch and dataset schedule.
- Each dataset can use a different per-GPU microbatch on the same accelerator.
- Each accelerator-memory class can use a different microbatch for the same task and dataset.
- The effective event count per optimizer update exactly equals `global_batch_size`.
- Accumulated gradients match an equivalent unaccumulated batch within numerical tolerance.
- LR scheduling, validation frequency, checkpointing, and resume operate in optimizer-step units.
- Invalid or incomplete batch plans fail during scenario resolution, before allocating a training job.

Dynamic token-budget batching and mixing multiple dataset groups inside one optimizer update can be considered later; the first implementation should use explicit static per-dataset microbatches.
